cs/src/core/Index/FASTER/FASTER.cs

```diff
@@ -319,6 +319,108 @@ internal Status ContextRead(ref Key key, ref Input input, ref Output output, Con
             return status;
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal ValueTask<(Status, Output)> ReadAsync(
+            ClientSession<Key, Value, Input, Output, Context, Functions> clientSession,
+            ref Key key, ref Input input, Context context = default, CancellationToken token = default)
+        {
+            var pcontext = default(PendingContext);
+            Output output = default;
+
+            var nextSerialNum = clientSession.ctx.serialNum + 1;
+
+            OperationStatus internalStatus;
+
+            if (clientSession.SupportAsync) clientSession.UnsafeResumeThread();
+            try
+            {
+                do
+                {
```

Comment:
This loop is not identical in behavior to the original code. For example, I notice that InternalRefresh would not be called in this tight loop, resulting in never getting out of CPR. The debug asserts for version correctness are also skipped. We need to ensure that all code paths are functionally identical to the older case.

Why not use an async equivalent of HandleOperationStatus? This keeps the main FasterKV.ReadAsync method small so it can be inlined, leaving the complexity of uncommon cases to a sub-method.